Features/feature skipped question

Amnish04 opened the pull request:
<!--
Thank you for opening this pull request! Your help is much appreciated.
Please choose the PR Type and give as many details as possible about your PR.
-->

Fixes #81 <!-- link to issue if one exists -->

- [x] I have read and agree to the [Code of Conduct](https://github.com/eduardconstantin/azure-fundamentals/blob/main/CODE_OF_CONDUCT.md), and have followed the [Contributing guidelines](https://github.com/eduardconstantin/azure-fundamentals/blob/main/CONTRIBUTING.md).

## Pull Request Type

<!-- Please delete options that are not relevant -->

- [x] Bug fix (please link related issues if applicable)
- [x] New feature
- [ ] Question update
- [ ] Other

## Summary

<!--
 Detailed explanation for the changes of your pull request
-->
@eduardconstantin  
Just finished with my implementation of requested functionality. Here is what I added:
- Added the implementation of a basic FIFO queue to be used in the actual feature.
- Any skipped questions are added to a `skippedQueue` saved in component state.
- Once the user goes over all the questions, the state is marked as `allQuestionsTouched`, and the user is brought to the first skipped question.
- Now every time user clicks next, he/she is taked to the next skipped question while the current question is removed from queue if answered.
- If the user decides to skip an already `skipped` question, they are brought back to it once they complete their current cycle of skipped questions.
- Fixed an **existing bug**, where `next button` was disabled in any scenario. Now, the next button only gets enabled once the user selects an option.

Hope you find these changes useful. Please let me know if something needs to be changed.

Thanks!

eduardconstantin commented on utils/Queue.ts:
Could you replace the class with function?

Amnish04 replied:
Just finished with latest imlementation.

eduardconstantin commented:
I also found a bug, if you select one option and then skip the question, when the question appears again the option remains selected.
![image](https://github.com/eduardconstantin/azure-fundamentals/assets/9199763/56b4f32d-9f8c-49da-ae48-328ce84576dd)

You can fix this or create an issue for it.

Amnish04 commented:
@eduardconstantin Just finished with all the changes requested.
1. Switched from a class to a functional style for `Queue` data structure.
2. Fixed the bug where the skipped questions preserve any selection. (I feel like that was a feature instead of a bug)
3. Resolved merge conflicts with latest changes in dev.

Hope it addresses the issues.

masterHAWK99 commented on pages/exam/index.tsx:
```diff
@@ -67,16 +70,41 @@ const Exam: NextPage = () => {
+    skippedQuestions.enqueue(questionNo);
+
+    // Unset any selected answer
+    setAnswers(prevState => {
+      if (prevState[questionNo]) delete prevState[questionNo];
+      return prevState;
+    });
```
You shouldn't modify the original object from state, instead, create a copy of this object, delete key, and finally update the state with a new object.

Amnish04 replied:
Fixed in the latest commit. Thanks for pointing out

masterHAWK99 commented on components/ExamQuizForm.tsx:
```diff
@@ -58,14 +62,21 @@ const ExamQuizForm: React.FC<Props> = ({
+        type="button" 
+        intent="primary" 
+        size="medium"
+        onClick={() => {
+          reset(); handleSkipQuestion(currentQuestionIndex);
+        }}
```
Please reformat your code manually, as you may have forgotten to run `npm run prepare`, causing it to not format while being committed.

Amnish04 replied:
You're right, I skipped that part. Should be fixed in the latest commit

masterHAWK99 commented on components/ExamQuizForm.tsx:
```diff
@@ -8,18 +8,22 @@ type Props = {
+  isQuestionAnswered,
 }) => {
   const { register, handleSubmit, reset } = useForm();
   const [lastIndex, setLastIndex] = useState<number>(0);
```
I meant `const [lastIndex, setLastIndex] = useState<number>(0);`

Amnish04 replied:
I see. To be honest, I am not sure why it was there in the first place. So I wouldn't prefer removing something I didn't add.

[Do not remove a fence until you know why it was put up in the first place.](https://fs.blog/chestertons-fence/)

eduardconstantin replied:
@masterHAWK99 this is yours
![image](https://github.com/eduardconstantin/azure-fundamentals/assets/9199763/5f1ef8d5-e6a4-47d8-8f76-19c16ae0727a)

eduardconstantin commented:
> I believe that in the top left corner, there should be a display of the number of answered questions. Currently, when we skip some questions and the test ends, it shows the index number of the skipped question. For example, if I skipped the 3rd question and answered the remaining questions, when the test ended, the display in the top left corner showed "3/30" instead of "29/30". This may cause confusion.
> 
> What do you think @eduardconstantin ?

Either that or the skipped questions should not be counted. I tested with 4 questions, I skipped 2 and answered 2, then the counter got resetted, I answered the 2 skipped questions and got 2/4 question answered, which is wrong, I answered all 4.
![image](https://github.com/eduardconstantin/azure-fundamentals/assets/9199763/99e82a35-9284-4955-863d-ec3818999d2a)

Amnish04 commented:
> > I believe that in the top left corner, there should be a display of the number of answered questions. Currently, when we skip some questions and the test ends, it shows the index number of the skipped question. For example, if I skipped the 3rd question and answered the remaining questions, when the test ended, the display in the top left corner showed "3/30" instead of "29/30". This may cause confusion.
> > What do you think @eduardconstantin ?
> 
> Either that or the skipped questions should not be counted. I tested with 4 questions, I skipped 2 and answered 2, then the counter got resetted, I answered the 2 skipped questions and got 2/4 question answered, which is wrong, I answered all 4. ![image](https://user-images.githubusercontent.com/9199763/277404448-99e82a35-9284-4955-863d-ec3818999d2a.png)

@eduardconstantin 
Yes, currently it shows the index of the skipped question you're on.
In other words, you're on ith question out of n questions (i/n).

One thing we can do is update the questions counter when all skipped questions have been answered, like in normal case with no skipped questions.
That would look like `n/n`.

Amnish04 commented:
That is due to the way application has been set up. It is using an array of questions as the source `data.randomQuestions`, and updates the current index to switch to a question.
The counter above, based on the existing logic, does not represent the number of answered questions but what question we're currently at. And when you're done with the inital cycle of questions, you're sequentially taken back to the skipped questions. For example, you skipped question x, then y.

Amnish04 commented:
@eduardconstantin Also, to address the issue of knowing current question number, I have added that before a question's text.

![image](https://github.com/eduardconstantin/azure-fundamentals/assets/78865303/373bc1d5-6aa2-433a-b201-ae08430fdba6)

eduardconstantin commented:
> @eduardconstantin Also, to address the issue of knowing current question number, I have added that before a question's text.
> 
> ![image](https://user-images.githubusercontent.com/78865303/277418485-373bc1d5-6aa2-433a-b201-ae08430fdba6.png)
> 
> Just pushed

This is not a bad idea

Amnish04 commented:
> > @eduardconstantin Also, to address the issue of knowing current question number, I have added that before a question's text.
> > ![image](https://user-images.githubusercontent.com/78865303/277418485-373bc1d5-6aa2-433a-b201-ae08430fdba6.png)
> > Just pushed
> 
> This is not a bad idea

The changes are already in!

eduardconstantin commented:
I tested and it look much better!
One thing though, when I select an option it gets counted as an answered question, shouldn't be counted when I press the next question button?
I can merge it like this and you can create an issue for this bug or you can fix it in this PR, it's up to you.

Amnish04 commented:
> I tested and it look much better! One thing though, when I select an option it gets counted as an answered question, shouldn't be counted when I press the next question button? I can merge it like this and you can create an issue for this bug or you can fix it in this PR, it's up to you.

That is because the current logic is setting the answers object, whenever a user clicks on an option instead of next button click handler. I think it belongs to a separate issue, will open it for you.

Amnish04 commented:
Great, just one last thing. Could you also add `hacktoberfest-accepted` label to this PR. Thanks!

eduardconstantin commented:
I added the hacktoberfest label, it should count
